sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/ProgressReporter.scala

```diff
@@ -180,15 +180,59 @@ trait ProgressReporter extends Logging {
     currentStatus = currentStatus.copy(isTriggerActive = false)
   }
 
+  /**
+   * Extract statistics about stateful operators from the executed query plan.
+   * SPARK-19378: Still report stateOperator metrics even though no data was processed while
+   * reporting progress.
+   */
+  private def extractStateOperatorMetrics(hasNewData: Boolean): Seq[StateOperatorProgress] = {
+    if (lastExecution == null) return Nil
+    // lastExecution could belong to one of the previous triggers if `!hasNewData`.
+    // Walking the plan again should be inexpensive.
+    val stateNodes = lastExecution.executedPlan.collect {
+      case p if p.isInstanceOf[StateStoreSaveExec] => p
+    }
+    stateNodes.map { node =>
+      val numRowsUpdated = if (hasNewData) {
+        node.metrics.get("numUpdatedStateRows").map(_.value).getOrElse(0L)
+      } else {
+        0L
+      }
+      new StateOperatorProgress(
+        numRowsTotal = node.metrics.get("numTotalStateRows").map(_.value).getOrElse(0L),
+        numRowsUpdated = numRowsUpdated)
+    }
+  }
+
+  /**
+   * Extract statistics about event time from the executed query plan.
+   * SPARK-19378: Still report eventTime metrics even though no data was processed while
+   * reporting progress.
+   */
+  private def extractEventTimeStats(watermarkTs: Map[String, String]): Map[String, String] = {
```

Comment:
it does not make sense for this method to take this watermarkTs as a param. its not extracting event time states from watermark ts, its just appending it. Then why not just return empty map, and do the appending outside? Or do the extraction of watermark inside the function as well.